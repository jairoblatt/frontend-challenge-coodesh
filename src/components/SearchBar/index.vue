<template>
  <div class="search-bar">
    <input class="search-bar__input" type="text" placeholder="Search user:" @input="onInput" />
    <div class="search-bar__prepend">
      <Icon size="28">mdi-account</Icon>
    </div>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";

export default Vue.extend({
  props: {
    debounceDelay: {
      type: Number,
      default: 400,
    } as PropOptions<number>,
  },

  data: () => ({
    timeoutId: 0,
  }),

  methods: {
    onInput($event: Event) {
      clearTimeout(this.timeoutId);
      this.timeoutId = setTimeout(() => {
        this.$emit("input", ($event.target as HTMLInputElement)?.value);
      }, this.debounceDelay);
    },
  },
});
</script>

<style lang="postcss" scoped>
.search-bar {
  @apply w-full border border-gray-100 bg-white rounded-md pl-5 flex justify-between items-center overflow-hidden transition-colors;
}

.search-bar__input {
  @apply outline-none h-full pl-5 w-full text-light-onSurfacePrimary font-semibold;
}

.search-bar__prepend {
  @apply border border-gray-100 flex justify-center;
}

.search-bar__prepend >>> i {
  @apply text-light-onSurfacePrimary;
}

.search-bar:focus-within {
  @apply border-primary-light;
}

.search-bar:focus-within .search-bar__prepend {
  @apply bg-primary-light border-primary-light;
}

.search-bar:focus-within .search-bar__prepend >>> i {
  @apply text-white;
}
</style>
