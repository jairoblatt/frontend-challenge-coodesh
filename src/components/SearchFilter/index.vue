<template>
  <div v-clickOutside="clickOutsideHandle" class="search__filter">
    <span class="search__title">
      <slot />
    </span>

    <div @click="showOptions = !showOptions" class="filter__input">
      <span> {{ resolveName }} </span>
      <Btn> <Icon :icon="openedIcon" /> </Btn>
    </div>

    <transition name="fade-top-2x">
      <ul v-show="showOptions" class="filter__options">
        <li
          v-for="(option, key) in options"
          :key="`option-${key}`"
          :class="{ 'option--active': option.value === value }"
          @click="selectOption(option)"
        >
          {{ option.name }}
        </li>
      </ul>
    </transition>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";
import { clickOutside } from "@/directives";

interface IOption {
  value: number | string;
  name: string;
}

export default Vue.extend({
  props: {
    options: Array as PropType<IOption[]>,
    value: [String, Number] as PropType<string | number>,
  },

  directives: { clickOutside },

  data: () => ({
    showOptions: false,
  }),

  computed: {
    openedIcon(): string {
      return this.showOptions ? "mdi-chevron-up" : "mdi-chevron-down";
    },

    resolveName(): string {
      const { options, value } = this;
      let optionsSelected: IOption | undefined;

      if (Array.isArray(options) && value) {
        optionsSelected = options.find((option: IOption) => {
          return option.value === value;
        });
      }

      return optionsSelected?.name || "";
    },
  },

  methods: {
    selectOption(option: IOption) {
      if (this.value !== option.value) {
        this.$emit("input", option.value);
        this.showOptions = false;
      }
    },

    clickOutsideHandle() {
      this.showOptions = false;
    },
  },
});
</script>

<style lang="postcss" scoped>
.search__filter {
  @apply relative ml-2 h-full w-full;
}

@screen md {
  .search__filter {
    width: 12rem;
  }
}

.search__title {
  @apply text-light-onSurfacePrimary font-semibold text-base;
}

.filter__input {
  @apply bg-white flex cursor-pointer items-center justify-between rounded-lg mt-1 border border-gray-100;
}

.filter__input span {
  @apply outline-none text-light-onSurfacePrimary pl-3 whitespace-nowrap overflow-ellipsis;
}

.filter__input >>> i {
  @apply text-light-onSurfacePrimary;
}
.filter__options {
  @apply absolute cursor-pointer z-30 rounded-sm mt-1 border-gray-100 p-1 bg-white shadow-lg text-light-onSurfacePrimary w-full overflow-y-auto;

  max-height: 15rem;
}

.filter__options li {
  @apply p-2 border border-gray-100 hover:bg-gray-50  mt-1 rounded-lg transition-colors;
}

.option--active {
  @apply bg-gray-100;
}
</style>
