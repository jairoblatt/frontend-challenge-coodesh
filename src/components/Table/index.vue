<template>
  <table class="table__wrapper">
    <thead>
      <tr>
        <th>Name</th>
        <th>Gender</th>
        <th>Birth</th>
        <th>Actions</th>
      </tr>
    </thead>

    <transition-group tag="tbody" name="slide-down">
      <TableItem
        v-for="(user, index) in users"
        :key="`table-item-${index}`"
        :name="user.name"
        :gender="user.gender"
        :age="user.dob.age"
        :avatar="user.picture.thumbnail"
      >
        <Btn outlined small @click="redirectToUserPage(user)">
          View
        </Btn>
      </TableItem>
    </transition-group>
  </table>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";
import TableItem from "./TableItem.vue";
import { UserData } from "@/types/IUser";

export default Vue.extend({
  components: {
    TableItem,
  },

  props: {
    users: {
      type: Array,
      default: () => [],
    } as PropOptions<UserData[]>,
  },

  methods: {
    redirectToUserPage(user: UserData) {
      this.$router.push({
        name: "user",
        params: { username: user.login.username },
        query: this.$route.query,
      });
    },
  },
});
</script>

<style lang="postcss" scoped>
.table__wrapper {
  @apply rounded-sm border border-gray-100 overflow-hidden text-center bg-white border-collapse w-full;
}

.table__wrapper thead tr {
  @apply text-light-onSurfacePrimary;
}

.table__wrapper th {
  @apply uppercase font-bold;

  padding: 1rem 2rem;
  letter-spacing: 0.1rem;
  font-size: 0.7rem;
}

.table__wrapper tr:nth-child(even) {
  @apply bg-gray-50 bg-opacity-60;
}

.table__wrapper tbody tr {
  @apply border border-gray-100 border-solid;
}
</style>
