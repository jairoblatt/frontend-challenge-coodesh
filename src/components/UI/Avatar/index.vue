<template>
  <component :is="tag" class="avatar" :style="styles">
    <slot />
  </component>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Avatar",

  props: {
    tag: {
      type: String,
      default: "div",
    } as PropOptions<string>,
    // The prop size it's definedly with 'rem' css unit meansure
    size: {
      type: [String, Number],
      default: 2,
    } as PropOptions<string | number>,
    // The prop border it's definedly with 'px' css unit meansure
    border: [String, Number] as PropType<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        height: `${this.size}rem`,
        width: `${this.size}rem`,
        borderWidth: `${this.border || Number(this.size) / 2}px`,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.avatar {
  @apply rounded-full border-4 border-primary-base border-solid overflow-hidden;
}

.avatar img,
source {
  @apply min-w-full h-auto object-contain;
}

/* absolute -top-28 left-2/4 transform -translate-x-2/4 */

/* .user__avatar img,
source {
  height: 12rem;
  width: 12rem;
} */
</style>
