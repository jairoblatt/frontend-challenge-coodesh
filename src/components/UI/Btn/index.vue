<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
