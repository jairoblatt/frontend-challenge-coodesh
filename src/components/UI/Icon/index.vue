<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
