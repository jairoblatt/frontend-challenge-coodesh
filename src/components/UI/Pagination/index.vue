<template>
  <div class="pagination" :class="{ 'pagination--active': this.loading }">
    <transition name="slide-up">
      <Spinner v-show="loading" size="40" />
    </transition>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import { sleep } from "@/utils/index";

export default Vue.extend({
  props: {
    distance: {
      type: Number,
      default: 40,
    } as PropOptions<number>,

    delay: {
      type: Number,
      default: 1000,
    } as PropOptions<number>,

    loading: Boolean as PropType<boolean>,
    forceInit: Boolean as PropType<boolean>,
  },

  data: () => ({
    ready: true,
  }),

  mounted() {
    this.forceInit && this.onLoadMore();
    document.addEventListener("scroll", this.infiniteScroll);
  },

  beforeDestroy() {
    document.removeEventListener("scroll", this.infiniteScroll);
  },

  methods: {
    async infiniteScroll() {
      const scroll = document.documentElement.clientHeight + document.documentElement.scrollTop;

      if (scroll >= document.documentElement.scrollHeight - this.distance && this.ready) {
        this.ready = false;
        this.onLoadMore();

        await sleep(this.delay);
        this.ready = true;
      }
    },

    onLoadMore() {
      this.$emit("loadMore");
    },
  },
});
</script>

<style lang="postcss" scoped>
.pagination {
  @apply w-full fixed bottom-0 left-2/4 transform -translate-x-2/4;
}

.pagination--active {
  @apply transition-colors;

  height: 10rem;
  background: linear-gradient(0deg, rgba(199, 199, 199, 0.548) 0%, rgba(255, 255, 255, 0) 86%);
}

.slide-enter,
.slide-leave-to {
  opacity: 0;
  transform: translateY(100%);
}

.slide-enter-active,
.slide-leave-active {
  transition: opacity 0.2s, transform 0.2s ease-in-out;
}
</style>
