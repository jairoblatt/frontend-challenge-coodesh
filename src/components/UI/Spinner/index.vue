<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
