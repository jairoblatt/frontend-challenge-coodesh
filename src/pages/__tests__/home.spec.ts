import { mount, MountOptions, Wrapper, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import * as UsersModule from "@/store/modules/users";
import { MockUser } from "@/utils/tests";
import SearchBar from "@/components/SearchBar/index.vue";
import SearchFilter from "@/components/SearchFilter/index.vue";
import Table from "@/components/Table/index.vue";
import Pagination from "@/components/UI/Pagination/index.vue";
import Avatar from "@/components/UI/Avatar/index.vue";
import Btn from "@/components/UI/Btn/index.vue";
import Icon from "@/components/UI/Icon/index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";
import Home from "../index.vue";
import { clickOutside } from "@/directives";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.directive("clickOutside", clickOutside);
Object.entries({ SearchBar, SearchFilter, Table, Pagination, Avatar, Btn, Icon, Spinner }).forEach(
  ([componentName, component]) => {
    localVue.component(componentName, component);
  }
);

type Instance = InstanceType<typeof Home>;
let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

const spyMethod = (methodName: string) => {
  return jest.spyOn((Home as any).options.methods, methodName);
};

const spyOns = {
  fetchUser: spyMethod("fetchUser"),
};

const $router = {
  push: jest.fn(),
  replace: jest.fn(),
};

const $route = {
  query: {
    seed: "test",
    page: "1",
  },
};

describe("Home.vue ", () => {
  beforeEach(() => {
    const store = new Vuex.Store({
      modules: {
        users: {
          namespaced: true,
          ...UsersModule,
          state: {
            users: [MockUser],
          },
        },
      },
    });

    mountFunction = (options = {}) => {
      return mount(Home, {
        localVue,
        store,
        stubs: ["RouterView"],
        mocks: {
          $router,
          $route,
        },
        ...options,
      });
    };
  });

  afterEach(() => {
    $router.replace.mockClear();
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should call "fetchUser" method when its created', () => {
    expect(spyOns.fetchUser).toHaveBeenCalledWith($route.query);
  });

  it('Should set in the "filter" data the gender and nat query', () => {
    const query = { gender: "male", nat: "australian" };
    const wrapper: any = mountFunction({
      mocks: {
        $route: {
          query,
        },
      },
    });

    expect(wrapper.vm.filters.gender).toBe(query.gender);
    expect(wrapper.vm.filters.nat).toBe(query.nat);
  });

  it("Should replace route", () => {
    const wrapper = mountFunction();
    const fetchInfo = {
      seed: "another_seed",
      page: "2",
    };

    (wrapper.vm as any).$options.watch.usersFetchInfo.call(wrapper.vm, fetchInfo);

    expect($router.replace).toHaveBeenCalledWith({ query: fetchInfo });
  });

  it("Should not replace route when the query page/gender/nat it's the same of fetchInfo page/gender/nat", () => {
    const wrapper = mountFunction();
    (wrapper.vm as any).$options.watch.usersFetchInfo.call(wrapper.vm, { page: 1 });

    expect($router.replace).not.toHaveBeenCalled();
  });

  it('Should return undefined when the watcher "usersFetchInfo" dont have a truthy param', () => {
    const wrapper = mountFunction();
    const usersFetchInfo = (wrapper.vm as any).$options.watch.usersFetchInfo;

    expect(usersFetchInfo.call(wrapper.vm, null)).toBe(undefined);
  });
});
