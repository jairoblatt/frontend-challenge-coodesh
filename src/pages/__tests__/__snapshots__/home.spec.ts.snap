// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue  Should mount the component 1`] = `
"<section class=\\"users\\">
  <div class=\\"users__header\\">
    <div class=\\"users__search\\">
      <div class=\\"search-bar\\" value=\\"\\"><input type=\\"text\\" placeholder=\\"Search user:\\" class=\\"search-bar__input\\">
        <div class=\\"search-bar__prepend\\"><i class=\\"mdi mdi-account\\" style=\\"font-size: 28px;\\"></i></div>
      </div>
    </div>
    <div class=\\"users__filters\\">
      <div class=\\"search__filter\\"><span class=\\"search__title\\">
        Gender
      </span>
        <div class=\\"filter__input\\"><span> All </span> <button class=\\"btn\\">
            <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-chevron-down\\" style=\\"font-size: 14px;\\"></i>
          </button></div>
        <transition-stub name=\\"fade-top-2x\\">
          <ul class=\\"filter__options\\" style=\\"display: none;\\">
            <li class=\\"option--active\\">
              All
            </li>
            <li class=\\"\\">
              Female
            </li>
            <li class=\\"\\">
              Male
            </li>
          </ul>
        </transition-stub>
      </div>
      <div class=\\"search__filter\\" data-testid=\\"filter-nationality\\"><span class=\\"search__title\\">
        Nationalities
      </span>
        <div class=\\"filter__input\\"><span> All </span> <button class=\\"btn\\">
            <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-chevron-down\\" style=\\"font-size: 14px;\\"></i>
          </button></div>
        <transition-stub name=\\"fade-top-2x\\">
          <ul class=\\"filter__options\\" style=\\"display: none;\\">
            <li class=\\"option--active\\">
              All
            </li>
            <li class=\\"\\">
              Australian
            </li>
            <li class=\\"\\">
              Brazilian
            </li>
            <li class=\\"\\">
              Canadian
            </li>
            <li class=\\"\\">
              Switzerland
            </li>
            <li class=\\"\\">
              German
            </li>
            <li class=\\"\\">
              Danish
            </li>
            <li class=\\"\\">
              Spanish
            </li>
            <li class=\\"\\">
              Finnish
            </li>
            <li class=\\"\\">
              French
            </li>
            <li class=\\"\\">
              United Kingdom
            </li>
            <li class=\\"\\">
              Irish
            </li>
            <li class=\\"\\">
              Iraqi
            </li>
            <li class=\\"\\">
              Norwegian
            </li>
            <li class=\\"\\">
              Dutch
            </li>
            <li class=\\"\\">
              New Zealand
            </li>
            <li class=\\"\\">
              Turkey
            </li>
            <li class=\\"\\">
              American
            </li>
          </ul>
        </transition-stub>
      </div>
    </div>
  </div>
  <div class=\\"users__content\\">
    <table class=\\"table__wrapper\\">
      <thead>
        <tr>
          <th>Name</th>
          <th>Gender</th>
          <th>Birth</th>
          <th>Actions</th>
        </tr>
      </thead>
      <transition-group-stub tag=\\"tbody\\" name=\\"slide-down\\">
        <tr class=\\"table-item\\">
          <td class=\\"item__avatar\\">
            <div class=\\"avatar mr-2\\" style=\\"height: 2.1rem; width: 2.1rem; border-width: 1.5px;\\"><img src=\\"https://randomuser.me/api/portraits/thumb/women/6.jpg\\" alt=\\"Liva Christiansen avatar\\"></div>
            Liva Christiansen
          </td>
          <td>female</td>
          <td>51</td>
          <td><button class=\\"btn btn--small btn--outlined\\">
              <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div>
              View
            </button></td>
        </tr>
      </transition-group-stub>
    </table>
  </div>
  <div class=\\"pagination pagination--active\\">
    <transition-stub name=\\"slide-up\\">
      <div class=\\"spinner-container\\"><span class=\\"spinner\\" style=\\"min-height: 40px; min-width: 40px;\\"></span></div>
    </transition-stub>
  </div>
  <div class=\\"users__children\\">
    <transition-stub name=\\"zoom-fade\\">
      <routerview-stub></routerview-stub>
    </transition-stub>
  </div>
</section>"
`;
